### PR TITLE
gateway-api: fix hostname gateway addresses

### PR DESCRIPTION
**Pull request by frankbu**
**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svcname, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svcname)][obj.Namespace]
+					if svc.Attributes.ClusterExternalPorts != nil {
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:1457]: @howardjohn Is the right way to check if the service is not LoadBalancer type?

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: I am not sure I understand what we are trying to do here? ResolveGatewayInstances already partitions internal vs external addresses?

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: It's trying to only use internal addresses (hostnames) if there are no external addresses AND the service in not LoadBalancer-type. This was to avoid the problem, you mentioned to me a couple of weeks ago, where we currently set the hostname initially, but then replace it after a delay, when the LoadBalancer address is ready. It's also stripping the port from the hostnames before setting the gateway addresses.

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: I feel like we should have this logic inside ResolveGatewayInstances though and just change that as needed. Or at least use the same logic, which is not `ClusterExternalPorts` (although, probably ends up the same in practice)?

One thing I am not sure of  - using ClusterExternalPorts or ClusterExternalAddress may not handle the case where the LB is still `<provisioning>`

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: Like we may not actually store the fact that "This is aLoadBalancer service but we don't have an IP yet" anywhere in model.service

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: > Like we may not actually store the fact that "This is aLoadBalancer service but we don't have an IP yet" anywhere in model.service

Yes, that seems to be the case. If it's a LoadBalancer-type service that is still pending we will not have external addresses: https://github.com/istio/istio/blob/e8c809f026c18823bc57c1419b064b48158bb5ad/pilot/pkg/serviceregistry/kube/conversion.go#L141

I was hoping that external ports would be the differentiator, but maybe not?

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: Looks like ClusterExternalPorts is just for NodePort?

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: Ah right, I missed that - thought it was ClusterIP. So, I guess we have 2 options:

1. Add another field (e.g., LoadBalanceType bool).
2. Just strip the port and remove duplicates, but still set the hostname, for now.

IIRC, you mentioned that the conformance tests are soon going to be checking for gateway address to confirm ready, so if we go with option 2, we may fail the conformance test next time me update.

WDYT?

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1457]: (1) probably makes the most sense. Maybe am enum `Type`  under ServiceAttributes instead of a bool though?

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go]: updated, PTAL

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 2]
```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svchost, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svchost)][obj.Namespace]
+					if svc.Attributes.Type != corev1.ServiceTypeLoadBalancer {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:1457]: IMO its better to have ResolveGatewayInstances incapsulate this logic in some way. Its a bit messy to have the  internal logic leak into gateway code (and a bug cause by someone refactoring - GatewayContext is suppose to encapsulate the internals of push_context and thus must live in the push context package...)

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 3]
```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svchost, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svchost)][obj.Namespace]
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:1456]: obj.Namespace is the namespace of the Gateway which only coincidentally aligns with the namespace of the hostname. This is another thing pointing to moving the logic into ResolveGatewayInstances

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 4]
```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
...
+					if svc.Attributes.Type != corev1.ServiceTypeLoadBalancer {
+						// Add internal hostname if not LoadBalancer type service and not duplicate
+						found := false
+						for _, addr := range addressesToReport {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:1460]: seems like we should have a warning like "Assigned to service(s) %s, but address is pending: %s" around L1420

File: pilot/pkg/model/service.go [hunk 5]
```diff
@@ -30,6 +30,7 @@ import (
 
 	endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
 	"github.com/mitchellh/copystructure"
+	corev1 "k8s.io/api/core/v1"
```
Comment (howardjohn) [inline pilot/pkg/model/service.go:33]: this adds a new dependency on the previously k8s agnostic `model` package. I am surprised this doesn't actually fail a test, I would have thought it would bloat the pilot-agent binary but maybe not. Regardless, it would be good to absrtact this; we can just copy the 4 strings into this package instead of import

**Review by howardjohn — COMMENTED** [review]
LGTM, just some small comments

**Comment by frankbu** [comment]
Review comments have been addressed. @howardjohn PTAL

File: pilot/pkg/config/kube/gateway/context.go [hunk 6]
```diff
@@ -40,17 +42,23 @@ func NewGatewayContext(ps *model.PushContext) GatewayContext {
...
+// * Internal addresses (eg istio-ingressgateway.istio-system.svc.cluster.local:80).
+// * External addresses (eg 1.2.3.4), this comes from LoadBalancer services. There may be multiple in some cases (especially multi cluster).
+// * Pending addresses (eg istio-ingressgateway.istio-system.svc), are LoadBalancer-type services with pending external addresses.
 // * Warnings for references that could not be resolved. These are intended to be user facing.
-func (gc GatewayContext) ResolveGatewayInstances(namespace string, gwsvcs []string, servers []*networking.Server) (internal, external, warns []string) {
+func (gc GatewayContext) ResolveGatewayInstances(
+	namespace string,
+	gwsvcs []string,
+	servers []*networking.Server,
+) (internal, external, pending, warns []string) {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/context.go:54]: super nit: once we are at 4 args its hard to follow, maybe we can move it into a struct

**Review by howardjohn — APPROVED** [review]
